### PR TITLE
drivers: syscon: Set default for CONFIG_SYSCON to y

### PR DESCRIPTION
**Pull request by yperess**
Currently for large projects with many boards, using the syscon driver
is a bit of a pain. It has the convinience of using dt_compat_enabled
for CONFIG_GENERIC_SYSCON but that only works if CONFIG_SYSCON is
also enabled.

Set the default for CONFIG_SYSCON=y so that the driver can be enabled
by simply setting the devicetree node to `status = "okay";`

Signed-off-by: Yuval Peress <peress@google.com>

### CODE REVIEW

**Comment by yperess** [comment]
For reference, https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3177749 is the change that integrates syscon. If we don't have this default value for CONFIG_SYSCON then we would need to also make changes to all the boards under `zephyr/boards/arm` _AND_ make sure that every new board also has this in the `_defconfig`. It's not impossible, but seems that the driver is already gated by the compatible string so there's no need to have double gating.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
How about using simple changing the default for `CONFIG_SYSCON` to `y` in the `soc/<arch>/<soc>/Kconfig.defconfig.series` for the SoCs containing a SYSCON device?

**Review by carlocaione — CHANGES_REQUESTED** [review]
This is ugly. What about using something like
```
default y if $(dt_compat_enabled,$(DT_COMPAT_SYSCON))
```
or something like that

**Comment by yperess** [comment]
My initial approach was actually to collapse the two into a single config SYSCON (which in a way is very similar to what you said @carlocaione). It would remove `SYSCON_GENERIC` since it's basically redundant and move the default `dt_compat_enabled` up to the SYSCON config, it would look something like:

```
DT_COMPAT_SYSCON := syscon

menuconfig SYSCON
	bool "SYSCON (System Controller) drivers"
	default $(dt_compat_enabled,$(DT_COMPAT_SYSCON))
	help
	  SYSCON (System Controller) drivers. System controller node represents
	  a register region containing a set of miscellaneous registers. The
	  registers are not cohesive enough to represent as any specific type
	  of device. The typical use-case is for some other node's driver, or
	  platform-specific code, to acquire a reference to the syscon node and
	  extract information from there.
if SYSCON

module = SYSCON
module-str = syscon
source "subsys/logging/Kconfig.template.log_config"

config SYSCON_INIT_PRIORITY
```

Would that work for you? I was thinking it was too big of a change and decided to go with the minimal approach originally.

**Comment by henrikbrixandersen** [comment]
> Would that work for you? I was thinking it was too big of a change and decided to go with the minimal approach originally.

This will require changes if we bring in a SoC specific SYSCON driver in the future. I think we need to keep the "top-level" Kconfig for enabling SYSCON support and then have the drivers underneath that (as it is today).

**Comment by yperess** [comment]
> > Would that work for you? I was thinking it was too big of a change and decided to go with the minimal approach originally.
> 
> This will require changes if we bring in a SoC specific SYSCON driver in the future. I think we need to keep the "top-level" Kconfig for enabling SYSCON support and then have the drivers underneath that (as it is today).

I see what you're saying, I would argue that CONFIG_SYSCON then is obsolete. We could do something similar to the logging template where CONFIG_SYSCON is enabled automatically if CONFIG_SYSCON_GENERIC is enabled (eventually other implementations). Should I update the PR so it's easier to see?

**Comment by henrikbrixandersen** [comment]
> I see what you're saying, I would argue that CONFIG_SYSCON then is obsolete. We could do something similar to the logging template where CONFIG_SYSCON is enabled automatically if CONFIG_SYSCON_GENERIC is enabled (eventually other implementations). Should I update the PR so it's easier to see?

As far as I can tell, this would go against how all other driver (except maybe interrupt controllers, where we recently had issues due to this exact construct (no top-level gating)) handles Kconfig.

File: drivers/syscon/Kconfig [hunk 1]
```diff
@@ -3,33 +3,31 @@
...
+# SYSCON (System Controller) drivers. System controller node represents
+# a register region containing a set of miscellaneous registers. The
+# registers are not cohesive enough to represent as any specific type
+# of device. The typical use-case is for some other node's driver, or
+# platform-specific code, to acquire a reference to the syscon node and
+# extract information from there.
+
+DT_COMPAT_SYSCON_GENERIC := syscon
+
+config SYSCON_GENERIC
```
Comment (tejlmand) [inline drivers/syscon/Kconfig:15]: With this change, the syscon generic driver is suddenly present at the top-level drivers Kconfig and not in a submenu like most other drivers.

image

The drivers menu is already very big, so please get this into a submenu again.

File: drivers/syscon/Kconfig [hunk 2]
```diff
@@ -3,33 +3,31 @@
...
-# SYSCON options
-#
-menuconfig SYSCON
-	bool "SYSCON (System Controller) drivers"
+# SYSCON (System Controller) drivers. System controller node represents
+# a register region containing a set of miscellaneous registers. The
+# registers are not cohesive enough to represent as any specific type
+# of device. The typical use-case is for some other node's driver, or
+# platform-specific code, to acquire a reference to the syscon node and
+# extract information from there.
```
Comment (tejlmand) [inline drivers/syscon/Kconfig:11]: Why is this removed from the `help` text.

This means it's no longer visible inside `menuconfig`.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: drivers/syscon/Kconfig (hunk 2 above)
Comment (yperess) [reply drivers/syscon/Kconfig:11]: This loops me back to my original change for this PR... where all I did was set the default value of SYSCON to `y` and kept the menu the same. That kept the help description the same and the same menu structure. If you'd like me to go back to the original 1 line change I'd be very happy to do that. It made the most sense, but you can see from the comments there was a lot of opposition to it.

Comment (tejlmand) [reply drivers/syscon/Kconfig:11]: > but you can see from the comments there was a lot of opposition to it.

I don't see opposition to keep the old structure, quite the opposite, like this comment:
https://github.com/zephyrproject-rtos/zephyr/pull/38762#issuecomment-926546586
> I think we need to keep the "top-level" Kconfig for enabling SYSCON support and then have the drivers underneath that (as it is today).

as I read the comments, the opposition was on how `SYSCON` was being enabled, as it goes against how other drivers do things.

Also make sure you're not re-introducing this: https://github.com/zephyrproject-rtos/zephyr/issues/38403
that is already now a mess to clean up.

Comment (yperess) [reply drivers/syscon/Kconfig]: OK, I reverted back to the original change + a change to address what you mentioned and avoid creating a `zephyr_library` unless the right Kconfig was enabled. The opposition was that this was messy and should be explicitly enabled. My take is that developers already need to add a devicetree node and set `status = "okay"`. This is explicit enough. Having `CONFIG_SYSCON=y` by default doesn't affect build time/size at all.

**Comment by yperess** [comment]
Ping, I went ahead with the original suggestion of defaulting the CONFIG_SYSCON the same as CONFIG_GENERIC_SYSCON. Thoughts?

File: drivers/syscon/CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-zephyr_library()
-zephyr_library_sources_ifdef(CONFIG_SYSCON          syscon.c)
+if(DEFINED CONFIG_SYSCON_GENERIC)
```
Comment (tejlmand) [inline drivers/syscon/CMakeLists.txt:3]: I believe this is no longer needed, as we already have this:
https://github.com/zephyrproject-rtos/zephyr/blob/ef0d955758e549f39d97b07089dedc4bdd05bc0e/drivers/CMakeLists.txt#L61

and you are now defaulting the `CONFIG_SYSCON` based on devicetree.

It is ok for a user enabling `CONFIG_SYSCON` but not adding any drivers to the library to see the warning, as it actually means there is a mismatch in the config settings.
```suggestion
```

**Review by tejlmand — COMMENTED** [review]
Make sense to default the syscon drivers menu based on the devicetree settings.
Downstream users with custom drivers can do similar, and use `zephyr_library_amend()` to place a custom driver in the lib.

Just a minor nit.

**Review by jackrosenthal — APPROVED** [review]

File: drivers/syscon/CMakeLists.txt (hunk 3 above)
Comment (yperess) [reply drivers/syscon/CMakeLists.txt:3]: Done

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
My initial comment still stands:
 
> How about using simple changing the default for `CONFIG_SYSCON` to `y` in the `soc/<arch>/<soc>/Kconfig.defconfig.series` for the SoCs containing a SYSCON device?

Why does this need to be handled differently than all other drivers classes?

**Comment by tejlmand** [comment]
> > How about using simple changing the default for `CONFIG_SYSCON` to `y` in the `soc/<arch>/<soc>/Kconfig.defconfig.series` for the SoCs containing a SYSCON device?
> 
> Why does this need to be handled differently than all other drivers classes?

But wouldn't that result in the same issue that @mbolivar-nordic  tries to cleanup here ?
https://github.com/zephyrproject-rtos/zephyr/pull/38510

Having a `config SYSCON\r default y` in board series Kconfig will result in one of those two things:
- The driver being enabled always for that board, even when not needed
   see #30694 
- `SYSCON` is enabled, but no driver is placed in that lib
    resulting in: #38403

Maybe we need to rethink how we enable drivers in general and where we place them in libraries.
Does each driver needs its own library if that library contains only a single file ?

**Comment by yperess** [comment]
> While I agree with @henrikbrixandersen I wanted to point out that something similar is already used in zephyr, see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/Kconfig.stm32#L7

I think there's precedence for most different ways to enable things. I would propose that we create an issue to discuss this and once there are several comments on it we can bring it up at the weekly dev meeting?

**Comment by yperess** [comment]
> > I think there's precedence for most different ways to enable things. I would propose that we create an issue to discuss this and once there are several comments on it we can bring it up at the weekly dev meeting?
> 
> Feel free to bring this to the next dev-meeting. Add a label for that in this case.

Done
